Add Shelley `TransactionLayer`

Anviking opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1672 

# Overview

- [x] Added `toCardano` conversion functions interacting with `cardano-api`
- [x] Fixed mistakes in existing conversion functions
- [x] Implemented Shelley `TransactionLayer` and enabled `Wallet` and `Transaction` integration tests.
- [x] Property test making sure estimateSize never underestimates
# Comments

Main commit: 4f6ea17

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -74,6 +105,64 @@ newTransactionLayer _proxy _protocolMagic = TransactionLayer
+        -- without converting from @SlotId@.
+        let epochLength = EpochLength 1215
+
+        let Cardano.TxUnsignedShelley unsigned = Cardano.buildShelleyTransaction
```
Would be nicer to use [`signTransaction`](https://github.com/input-output-hk/cardano-node/blob/c9ce637a260f415c8af21d75ddc6a1bb2efac23b/cardano-api/src/Cardano/Api.hs#L482-L496).

But I thought the way cardano-crypto carries an _encrypted_ XPrv and a password to the very end was intentional for… some security reasons? To construct `SignKeyDSIGN` we'd need to provide the unencrypted key.

KtorZ replied:
>  was intentional for… some security reasons

Partly. Having key encrypted in-memory was an explicit request from the previous development team which was several times depicted as useless by cryptographers. What matters is storing them encrypted (cf the wallet-chain-libs rust implementation). We still use cardano-crypto for lack of alternative.

KtorZ committed:
Faucet: regenerate shelley genesis UTxO with payment address & blake2b_256 hashes

Faucet: Re-generate shelley faucets and tweak generation script

1. Faucet wallets should not be random!
2. Format needs to be adjusted slightly

Regenerate faucet wallets with correct amount of funds

Regenerate faucet wallets

Tweak genesis

Anviking committed:
Aux: Test that ledger can deserialise faucet addresses

Anviking committed:
Fix: toSealed should serialize the consensus tx to match toGenTx

Anviking committed:
Fix: Retrieve txParams from genesis

Anviking committed:
Fix: use SL.serializeAddr instead of serialize'

Without this the addresses from the genesis block ended up a few bytes
too long.

This commit also re-uses fromShelleyAddress, instead of duplicating the
implementation.

Anviking committed:
Faucet: rewrite faucets in core-integration and cleanup shelley genesis yaml

Anviking committed:
Rework the format for genesis txouts

Anviking committed:
Aux: remove un-used mainnetBlockchainParameters in shelley

Anviking committed:
Add shelley transaction layer

Enable Shelley Transaction API

Enable Shelley listAddresses

Anviking committed:
Fix: Add Encode,DecodeAddr & Tx conversion functions

Anviking committed:
Take current tip when constructing txs in `TransactionLayer`

Needed for Shelley.

Anviking committed:
Fix mixup of minfeeA and minfeeB

Though the size estimation could use another look.

Anviking committed:
Hack: Swap mainnet and testnet network values

The FF testnet is marked as mainnet. This change was the easiest way to
support the FF.

Anviking committed:
Fixup: use correct passphrase in integration test

Why was it passing for jormungandr previously?

Anviking committed:
Aux: Add start.sh command for running shelley self-node

Useful for playing around

Anviking committed:
Make "TRANS_CREATE_04 - Not enough money" pass more easily

To make it work on shelley, without relying on that the minFee is
accurate.

Anviking committed:
fix weeder

Anviking committed:
Test that shelley tx estimateSize never underestimates

And tweak calculation

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -305,20 +305,20 @@ instance DelegationAddress ('Testnet pm) ShelleyKey where
       where
         base = 0
-        networkId = 0
+        networkId = 1 -- NOTE: Should be 0. Hack to support FF.
```
site-note: this should be `FIXME`, not `NOTE`. FIXME is used for code that will require change otherwise the software won't work. NOTE is for giving precision. (and TODO is more of a nice-to-have, but the software will work without).

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -15,92 +14,166 @@
+        let certs = []
+
+        -- TODO: The SlotId-SlotNo conversion based on epoch length would not
+        -- work if the epoch length changed in a hard fork.
```
Although, the transaction itself would be invalid before the hard-fork. True that, subsequent hardfork after shelley may be a risk but for `Byron` -> `Shelley`, there's no problem here if we know the new epoch length prior to the hardfork.

Anviking replied:
```
era  Byron | Shelley
el       3 | 2
epoch   0 1 2 3
=========================================
slot 0 |0 3 6 8
  V  1 |1 4 7 9
     2 |2 5
```

If we are here given a correct `SlotId 3 1` at `SlotNo 9` (which arguably would be unlikely, since there's a similar conversion to create the SlotId)

The we cannot simply convert back to `SlotNo` using the genesis epoch length (here 3):

```
(SlotNo 3*3+1) == (SlotNo 10)
But the correct answer is: SlotNo 9
```

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -15,92 +14,166 @@
+
+        -- TODO: The SlotId-SlotNo conversion based on epoch length would not
+        -- work if the epoch length changed in a hard fork.
+        let timeToLive = (toSlotNo epochLength slot) + 7200
```
7200 ? Avoid magic numbers, make it a named constant and explain the rationale of choosing `7200` (ultimately, we want that configurable via the API). Can you make sure to let Alex know about this?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -15,92 +14,166 @@
+                sig = SignedDSIGN $ fromMaybe (error "error converting signatures")
+                        $ rawDeserialiseSigDSIGN
+                        $ k `sign` body
+                vKey = SL.VKey $ VerKeyEd25519DSIGN $  throwCryptoError $ Ed25519.publicKey $ BS.take 32 $ CC.unXPub $ CC.toXPub $ getRawKey $ fst k
```
`BS.take 32 $ CC.unXPub $ CC.toXPub`

Use : `xpubPublicKey` instead from `Cardano.Address.Derivation`

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction/Size.hs:
```diff
@@ -39,26 +37,21 @@ class MinSizeOf (t :: *) (n :: NetworkDiscriminant) (k :: Depth -> * -> *) where
+    -- Maybe that value wasn't taking delegation keys into account?
+    --
+    -- Regardless:
+    -- TODO: Implement properly.
```
The length of the output will actually depend on the value and the type of addresses used. We can assume worst-case scenario which are currently 65 bytes for addresses indeed. 32 bytes for just the value sounds huge.

KtorZ reviewed (approved):
Nice work. I think we can go with this first big chunk and do improvements / additions on top.

KtorZ committed:
rewrite mkStdTx for readability & future maintenance

- split functions out so that they read better and will likely be re-used for implementing other parts of the transaction layer.
- avoid unnecessary identifiers declaration. Too many identifiers in a single small context don't make things more readable.
- changed a few types to avoid repeating the same unwrapping operation
- named some in-place calculation (like realFee)
- used forM + pattern-matching instead of mapM and pointfree style. Pointfree is usually not readable when it's in a context with more than one line

KtorZ committed:
sort dependencies alphabetically in shelley .cabal

Anviking committed:
Fixup: Use more ada in test to compensate for fee

Anviking committed:
Fixup: make jormungandr compile again

Anviking committed:
Explain rationale about the hard-coded TTL value

Anviking committed:
Fixup: Label network magic NOTEs as FIXMEs

KtorZ committed:
implement transaction size estimation & revise existing tests

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -131,10 +132,10 @@ newTransactionLayer _proxy _protocolMagic epochLength = TransactionLayer
       where
         dummyOutput :: Coin -> TxOut
-        dummyOutput = TxOut $ Address "\130\ACKX \217\NULYA\205=\179d\160d\203\232\SYN\138C\SOH\\MGO\141\201\228\161\138\146\136\179\235M\217v"
+        dummyOutput = TxOut $ Address $ BS.replicate 65 0
```
65 should be `maxSizeOf @Address`

KtorZ committed:
use resolved input in integration test fee estimator

KtorZ committed:
oversight: fix temporary directory prefix for shelley launcher

KtorZ committed:
use 'txsize' from cardano-ledger-specs for correct size calculation

The fee calculation isn't actually made on the transaction size in bytes, but rather, on an abstract representation of the transaction.
This is detailed in the ledger specifications and implemented in the cardano-ledger-specs module. So, we might as well re-use this here.

KtorZ committed:
increase transaction max size in genesis file

KtorZ committed:
disable some integration tests that aren't possible on Shelley

KtorZ committed:
clean up byron AND shelley wallets too between integration scenario...

KtorZ committed:
activate some non-implemented endpoint from the Shelley server

KtorZ committed:
replace hard-coded (and wrong) address in spec with valid address generated on the fly

KtorZ committed:
disable not-yet-covered scenario for migrating big byron wallets

KtorZ committed:
make scenario polling network information more reliable

By not depending on a pass to the next epoch, which can be quite long on the cardano-node.

Commit:
Regenerate nix
